<a href="https://colab.research.google.com/github/miguelgutierrez/03MAIR---Algoritmos-de-Optimizacion---2019/blob/master/Miguel_Gutierrez_Seminario_Algoritmos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos:   **Miguel Gutiérrez**<br>
Url: https://colab.research.google.com/drive/1Kfe0xTO1kuQHZRYj2y89OKpqoj4Ol7E_<br>

Problema: **Sesiones de doblaje** <br>


Se precisa coordinar el doblaje de una película. Los actores del doblaje deben coincidir en las tomas en las que sus personajes aparecen juntos en las diferentes tomas. Los actores de doblaje cobran todos la misma cantidad por cada día que deben desplazarse hasta el estudio de grabación independientemente del número de tomas que se graben. No es posible grabar más de 6 tomas por día. El objetivo es planificar las sesiones por día de manera que el gasto por los servicios de los actores de doblaje sea el menor posible. Los datos son:


*   Número de actores: **10**
*   Número de tomas: **30**
*   Actores/Tomas: https://bit.ly/36D8IuK  
(1 indica que el actor participa en la toma 0 en caso contrario)


(*) La respuesta es obligatoria





                                        

( * ) ***¿Cuantas posibilidades hay sin tener en cuenta las restricciones?***<br>

***¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.*** 




**Respuesta**<br>
**Sin restricciones**
Si quito la restricción de 6 tomas/día, significa que puedo hacer todas las tomas el mismo día. <br>Por lo tanto, habría 1 solución: realizar todas las tomas el mismo día. <br>Coste: 10 actores x coste del actor de doblaje.

**Con restricciones**<br>
(n fact r) * (n-r fact r) * ... (5 veces) <br>
(30 fact 6) * (24 fact 6) * (18 fact 6) * (12 fact 6) * (6 fact 6)<br>

Total posibles soluciones:  1370874167589326400

**Modelo para el espacio de soluciones**<br>
( * ) ***¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguméntalo)***


**Respuesta**

**tomas**:<br>
 
 Matriz con la información de las tomas y actores que intervienen (30 x 10):

*   Filas son las tomas: fila 0 -> Toma 1
*   Columnas: Columna 0 -> Actor 1<br>
    valor: 0/1 no trabaja/trabaja el actor en esa toma


**solucion**:<br>
 Lista de días, donde cada elemento es un set de tomas.
 Es una lista porque tiene que ser ordenada, representando:<br>
 
*   elemento 0 -> día 1
*   elemento 1 -> día 2
*   elemento n-1 -> día n

Cada elemento es un **set** de las tomas del día, ya que no necesitamos que sean ordenadas y además debe ser un conjunto de elementos (las tomas no pueden estar repetidas)

P.e.: [{1, 2, 3, 4, 5, 6}, {7, 8, 9, 10, 11, 12}, {13, 14, 15, 16, 17, 18}, {19, 20, 21, 22, 23, 24}, {25, 26, 27, 28, 29, 30}]


**tomas pendientes**:<br>
  Set de tomas (no necesito que estén ordenadas y sí sean elementos únicos), donde voy quitando las que se utilizan en cada día, para gestionar las tomas que se van a realizar el día siguiente

Inicialmente pensé en una matriz pero lo descarté porque no representaba los elementos y acciones que necesitaba para resolver el problema

**árbol/nodo raiz**:
  * lista de nodos y hojas
  * *hoja*: día (set de las tomas realizadas ese día)
  * *nodo*: lista de nodos y hojas



**Según el modelo para el espacio de soluciones**<br>
(*)¿Cual es la función objetivo?

(*)¿Es un problema de maximización o minimización?

**Respuesta**<br>
Mínimo de la suma del coste de las tomas de cada día

Es un problema de minimización

Diseña un algoritmo para resolver el problema por fuerza bruta

**Fuerza Bruta**

In [0]:
#Valores y funciones generales
TOMAS = [
[1,1,1,1,1,0,0,0,0,0],
[0,0,1,1,1,0,0,0,0,0],
[0,1,0,0,1,0,1,0,0,0],
[1,1,0,0,0,0,1,1,0,0],
[0,1,0,1,0,0,0,1,0,0],
[1,1,0,1,1,0,0,0,0,0],
[1,1,0,1,1,0,0,0,0,0],
[1,1,0,0,0,1,0,0,0,0],
[1,1,0,1,0,0,0,0,0,0],
[1,1,0,0,0,1,0,0,1,0],
[1,1,1,0,1,0,0,1,0,0],
[1,1,1,1,0,1,0,0,0,0],
[1,0,0,1,1,0,0,0,0,0],
[1,0,1,0,0,1,0,0,0,0],
[1,1,0,0,0,0,1,0,0,0],
[0,0,0,1,0,0,0,0,0,1],
[1,0,1,0,0,0,0,0,0,0],
[0,0,1,0,0,1,0,0,0,0],
[1,0,1,0,0,0,0,0,0,0],
[1,0,1,1,1,0,0,0,0,0],
[0,0,0,0,0,1,0,1,0,0],
[1,1,1,1,0,0,0,0,0,0],
[1,0,1,0,0,0,0,0,0,0],
[0,0,1,0,0,1,0,0,0,0],
[1,1,0,1,0,0,0,0,0,1],
[1,0,1,0,1,0,0,0,1,0],
[0,0,0,1,1,0,0,0,0,0],
[1,0,0,1,0,0,0,0,0,0],
[1,0,0,0,1,1,0,0,0,0],
[1,0,0,1,0,0,0,0,0,0]
]
max_tomas_dia = 6

In [0]:
#funciones generales
#toma: fila de valores 0,1 que indican si el actor en dicha posición trabaja en dicha toma
#actores de una toma: devuelve una lista con los actores 1,...,10
def actores_de_toma(toma):
  actores = []
  for index in range(len(toma)):
    if toma[index] == 1:
      actores.append(index+1)
  return actores
#solucion: tupla de sets de tomas
#devuelve el coste de la solución
def coste_de_solucion(TOMAS,solucion,coste_actor):
  coste=0
  for tomas_dia in solucion:
    coste = coste + coste_de_tomas_mismo_dia(TOMAS,tomas_dia,coste_actor)
    #print(tomas_dia)
  return coste
#tomas_dia: set de las tomas de un día (1,...,n)
#se forma un set de actores con las tomas. El total será el coste
#devuelve el coste: en este caso suponemos que todos los actores cobran lo mismo
def coste_de_tomas_mismo_dia(TOMAS,tomas_dia,coste_actor):
  actores = set({})
  for toma in tomas_dia:
    actores.update(actores_de_toma(TOMAS[toma-1]))    
  return len(actores) * coste_actor

#TEST
#print(coste_de_tomas_mismo_dia(TOMAS,{1,2,3,4,5,6},1))
#solucion = list()
#solucion.append({1,2,3,4,5,6})
#solucion.append({8, 10, 12, 13, 15, 17})
#solucion.append({7, 9, 11, 14, 16, 18})
#print(coste_de_solucion(TOMAS,solucion,1))

In [3]:
import itertools
from itertools import permutations
import math

def print_complejidad(num_tomas,max_tomas_dia):
    count = 1
    total_dias = math.ceil(num_tomas/max_tomas_dia)
    for dia in range(total_dias):
      count = count * len(list(itertools.combinations(range(num_tomas-dia*max_tomas_dia),max_tomas_dia)))
    print("Total posibles soluciones: ", count)

#cada día hacemos 6 tomas que es lo máximo que nos permiten
#día 0: recorremos todas las combinaciones de (total tomas, max por día) -> (T,MaxD)
#día 1: recorremos todas las combinaciones de (T-MaxD, MaxD)
#día d: si T>MaxD*d, recorremos todas las combinaciones de (T-MaxD*d, MaxD)
#posibles soluciones es O(n!) siendo n las tomas ya que son: (combinaciones (n 6): día1)*(combinaciones (n-6,6): día2)*(combinaciones (n-2*6, 6) día3)
#mejor_solucion: tupla de n días, donde cada valor es un set de tomas. El valor d de la tupla representa el set de tomas realizado el día d

#variables globales
mejor_solucion = list()
coste_mejor_solucion = 10e10

def fuerza_bruta(TOMAS):
  num_tomas = len(TOMAS)
  num_actores = len(TOMAS[0])
  print("Número tomas: ", str(num_tomas), " Número de actores:", str(num_actores))
  
  tomas_totales = set(range(1,len(TOMAS)+1)) 

  print_complejidad(num_tomas,max_tomas_dia) 

  solucion_parcial = list()
  tomas_pendientes = set(tomas_totales)
  
  calcula_solucion_parcial(tomas_totales,tomas_pendientes,solucion_parcial,max_tomas_dia)

  print("Mejor solucion:", mejor_solucion)
  print("coste mejor solucion:", coste_de_solucion(TOMAS,mejor_solucion,1))

#función recursiva, calcula soluciones para los días: d1,.., dn
def calcula_solucion_parcial(tomas_totales, tomas_pendientes, solucion_parcial, max_tomas_dia):
  global mejor_solucion, coste_mejor_solucion

  for s in list(itertools.combinations(tomas_pendientes,max_tomas_dia)):
    tomas_pendientes1 = set(tomas_pendientes)
    solucion_parcial1 = list(solucion_parcial)
    solucion_parcial1.append(set(s))

    for parcial in solucion_parcial1:
      #print("parcial:",parcial)
      tomas_pendientes2 = tomas_pendientes - set(parcial)

    tomas_pendientes2 = tomas_pendientes1 - set(s)
    #print("tomas_pendientes2:",tomas_pendientes2)

    if len(tomas_pendientes2) == 0:
      coste = coste_de_solucion(TOMAS,solucion_parcial1,1)
      #print("coste:", coste)
      if (coste < coste_mejor_solucion):
          mejor_solucion = solucion_parcial1 
          coste_mejor_solucion = coste         
          print("Mejor solucion parcial:", mejor_solucion)
          print("coste mejor solucion parcial:", coste_mejor_solucion)
    else:
      calcula_solucion_parcial(tomas_totales,tomas_pendientes2,solucion_parcial1,max_tomas_dia)

fuerza_bruta(TOMAS)

Número tomas:  30  Número de actores: 10
Total posibles soluciones:  1370874167589326400
Mejor solucion parcial: [{1, 2, 3, 4, 5, 6}, {7, 8, 9, 10, 11, 12}, {13, 14, 15, 16, 17, 18}, {19, 20, 21, 22, 23, 24}, {25, 26, 27, 28, 29, 30}]
coste mejor solucion parcial: 38
Mejor solucion parcial: [{1, 2, 3, 4, 5, 6}, {7, 8, 9, 10, 11, 12}, {13, 14, 15, 16, 17, 18}, {19, 20, 21, 22, 23, 25}, {24, 26, 27, 28, 29, 30}]
coste mejor solucion parcial: 37
Mejor solucion parcial: [{1, 2, 3, 4, 5, 6}, {7, 8, 9, 10, 11, 12}, {13, 14, 15, 16, 17, 18}, {19, 20, 21, 23, 24, 29}, {22, 25, 26, 27, 28, 30}]
coste mejor solucion parcial: 36
Mejor solucion parcial: [{1, 2, 3, 4, 5, 6}, {7, 8, 9, 10, 11, 12}, {13, 14, 15, 16, 17, 18}, {19, 21, 23, 24, 26, 29}, {20, 22, 25, 27, 28, 30}]
coste mejor solucion parcial: 35
Mejor solucion parcial: [{1, 2, 3, 4, 5, 6}, {7, 8, 9, 10, 11, 12}, {13, 14, 15, 16, 17, 25}, {18, 19, 21, 23, 24, 29}, {20, 22, 26, 27, 28, 30}]
coste mejor solucion parcial: 34
Mejor solucion p

KeyboardInterrupt: ignored

**Resultado Fuerza Bruta**<br>
Después de 60 minutos paro el proceso<br>
num_tomas:  30  num_actores: 10 num_dias:  5<br>
Mejor solucion parcial: [{1, 2, 3, 4, 5, 6}, {7, 8, 9, 10, 11, 12}, {13, 14, 20, 26, 27, 29}, {15, 16, 22, 25, 28, 30}, {17, 18, 19, 21, 23, 24}]<br>
coste mejor solucion parcial: 31<br>


***Calcula la complejidad del algoritmo por fuerza bruta***

**Respuesta**

(n fact r) * (n-r fact r) * ... (5 veces) <br>
(30 fact 6) * (24 fact 6) * (18 fact 6) * (12 fact 6) * (6 fact 6)<br>

**O(n!)**

Total posibles soluciones (30 tomas 6 tomas/día):  1370874167589326400

( * ) ***Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta***

Dado que el algoritmo por fuerza bruta no lo puedo utilizar dado el orden de complejidad, he optado por un algormitmo de ramificación y poda  con función de poda heurística <br>
(No puedo utilizar unas Cotas inforiores o superiores ya que el cálculo también es de un orden de magnitud similar al del algoritmo)

La función de poda heurística está basado en:<br>

*   La primera toma del día se coje la que tenga mayor número de actores (mayor coste). Puesto que tengo que hacerla cojo la mayor y a partir de ahí intento realizar las tomas aprovechando al máximo los actores que han tenido que ir
*   Las siguientes tomas las cojo utilizando una función de valoración (coste de oportunidad) calculando: <br>
los actores que ya estaban (ahorro)<br>
los actores que tienen que ir (coste)<br>
  valoracion = actores_tiene_ir-actores_ya_trabajando<br>
  cojo las tomas que menor valoración tengan (más ahorro tienen)<br>
  he puesto un máximo de 5 para evitar ramificaciones muy grandes, aunque con las pruebas realizadas nunca se llega<br>





Respuesta

In [0]:
import itertools
from itertools import permutations
import math

#variables globales
mejor_solucion = list()
coste_mejor_solucion = 10e10

#devuelve un valor: basado en el coste/ahorro que supone añadir una toma ese día o dejarla para el día siguiente
#Por ejemplo: si añado una toma que tiene 3 actores: 2 que ya están trabajando y 1 que no: -> -1 (ahorro de 1 actor)
#Por ejemplo: si añado una toma que tiene 2 actores: 2 que ya están trabajando: -> -2 (ahorro de 2 actores)
#Por ejemplo: si añado una toma que tiene 2 actores: 2 que ninguno están trabajando: -> 2 (tienen que ir 2 actores)
#tomas_dia: set con las tomas
#toma: número de la toma
def valoracion_annadir_toma(TOMAS,tomas_dia,numero_toma):
  actores_trabajando = set({})
  actores_toma = actores_de_toma(TOMAS[numero_toma-1])
  for toma in tomas_dia:
    actores_trabajando.update(actores_de_toma(TOMAS[toma-1]))    

  actores_ya_trabajando = actores_trabajando.intersection(actores_toma)

  actores_tiene_ir = set(actores_toma) - actores_ya_trabajando
  return (len(actores_tiene_ir) - len(actores_ya_trabajando))

#test
#tomas_dia = {1,3}
#toma = 6
#valoracion_annadir_toma(TOMAS,tomas_dia,toma)

#A partir de las tomas_dia (len >0) Genera los hijos que tengan la valoración menor 
#max_hijos: máximo para evitar supuestas ramificaciones excesivas
def crear_hijos(TOMAS,tomas,tomas_dia,max_hijos):
  HIJOS = []
  valoracion_minima = 10e10  

  tomas_pendientes = set(tomas) - tomas_dia
  #print("tomas_pendientes:", tomas_pendientes)
  for toma in tomas_pendientes: 
    valoracion = valoracion_annadir_toma(TOMAS,tomas_dia,toma)
    if valoracion < valoracion_minima:
      valoracion_minima = valoracion

  #print("valoracion_minima: ", valoracion_minima)
  for toma in tomas_pendientes: 
    valoracion = valoracion_annadir_toma(TOMAS,tomas_dia,toma)
    if valoracion_minima == valoracion and len(HIJOS)<max_hijos:
      HIJOS.append({toma})
  return HIJOS

#test
#tomas_totales = set(range(1,len(TOMAS)+1)) 
#print(crear_hijos(TOMAS,tomas_totales,{1,6,7,20,22},5))

#quita las tomas que no tengan ningún actor en común. Es decir, tomas que todos sus actores son nuevos ese día
#en ese caso, hacemos que vayan el día siguiente. Esto es factible, sólo en el caso, que no tengamos la restricción de realizar las tomas lo antes posible
def crear_hijos_si_tienen_actores_comunes(TOMAS,tomas,tomas_dia,max_hijos):
  HIJOS = crear_hijos(TOMAS,tomas, tomas_dia,max_hijos)
  actores_trabajando = set({})
  for toma in tomas_dia:
    actores_trabajando.update(actores_de_toma(TOMAS[toma-1]))    

  for hijo in HIJOS:
    actores_hijo = actores_de_toma(TOMAS[hijo-1])
    actores_ya_trabajando = actores_trabajando.intersection(actores_hijo)
    if len(actores_ya_trabajando) == 0:
      HIJOS.remove(hijo)
  return HIJOS

#test
#tomas_totales = { 11, 21} 
#print(crear_hijos(TOMAS,tomas_totales,{16},5))
#print(crear_hijos_si_tienen_actores_comunes(TOMAS,tomas_totales,{16},5))

#se utiliza para seleccionar el primer nodo del día
#todo: investigar otras funciones
def toma_con_mas_actores(TOMAS,tomas):
  max_actores = 0 
  toma_seleccionada = 0
  for toma in tomas: 
    count_actores = len(actores_de_toma(TOMAS[toma-1]))
    if count_actores > max_actores:
      toma_seleccionada = toma
      max_actores = count_actores    
  return toma_seleccionada

#test
#tomas_totales = set(range(1,len(TOMAS)+1)) 
#print(toma_con_mas_actores(TOMAS,tomas_totales))

In [14]:
import time

#Nodos del grafo  {1,3,5} números de las tomas. He incluído un máximo para evitar ramificaciones excesivas. 
#Nodos del grafo: ({dia0,(lista dia0)})
#   lista dia0: son las posibles tomas día 0 y para cada elemento se incluye (lista dia1), es decir tenemos un par: {dia, (lista dia)}
#   y así recursivamente hasta que no se pueden generar más días
#Se debería probar a cambiar dicho número

#NODOS: array de días con los nodos posibles de ese día
NODOS=[]
dia = 0

#lista de sets, nodo es un set
def existeNodoEnLista(nodo,lista):  
  for elemento in lista:
    if elemento == nodo:
      return True

  return False

def anadir_hijos(RAMA,HIJOS_ANTERIORES,tomas_pendientes,iteracion,max_tomas_dia):  
  HIJOS = []
  for nodo in HIJOS_ANTERIORES:
    for x in crear_hijos(TOMAS,tomas_pendientes,nodo,5):
      nuevo_nodo = nodo.union(x)
      #compruebo que no esté añadido
      if not existeNodoEnLista(nuevo_nodo,HIJOS):
        HIJOS.append(nuevo_nodo)

  #print("HIJOS FINAL", HIJOS)  
  if (iteracion+1) < max_tomas_dia:    
    anadir_hijos(RAMA,HIJOS,tomas_pendientes,iteracion+1,max_tomas_dia)
  else:
    RAMA.append(HIJOS)


def ramificacion_y_poda(TOMAS, tomas_pendientes):
  start_time = time.time()
  #nodo_inicial = set(toma_con_mas_actores(TOMAS, tomas_pendientes))  
  toma_mas_actores = toma_con_mas_actores(TOMAS, tomas_pendientes)
  #print("toma_mas_actores: ",toma_mas_actores)
  nodo_inicial = {toma_mas_actores}

  nodos_iniciales = []
  nodos_iniciales.append(nodo_inicial)
  rama = []
  anadir_hijos(rama,nodos_iniciales, tomas_pendientes,1,6)

  #todo: definir calcula_ramas, en función del día
  #bucle que lo independice de los días
  #dia1
  rama_dia_1 = calcula_ramas_dia1(rama[0],tomas_pendientes)
  #for elemento in rama_dia_1:
  #  print("elemento 1:", elemento)

  #dia2
  rama_dia_2 = calcula_ramas_dia2(rama_dia_1,tomas_pendientes)

  #dia3 
  rama_dia_3 = calcula_ramas_dia3(rama_dia_2,tomas_pendientes)

  #dia4
  rama_dia_4 = calcula_ramas_dia4(rama_dia_3,tomas_pendientes)

  #dia5, el último elemento es una lista con 1 sólo elemento
  soluciones = []
  for elemento in rama_dia_4:
    dia5 = elemento[4][0]
    soluciones.append([ elemento[0], elemento[1], elemento[2], elemento[3], dia5])

  print("Nivel complejidad. Total soluciones: ", len(soluciones))  
  solucion = calcula_mejor_solucion(TOMAS,soluciones)
  print("Duración: ", int((time.time() - start_time)), " segundos")

def calcula_ramas_dia1(rama,tomas_pendientes):
  rama_dia_1 = []
  for dia1 in rama:
    tomas_pendientes_dia1 = tomas_pendientes - dia1
    toma_mas_actores1 = toma_con_mas_actores(TOMAS, tomas_pendientes_dia1)

    nodo_inicial1 = {toma_mas_actores1}
    nodos_iniciales1 = []
    nodos_iniciales1.append(nodo_inicial1)
    rama1 = []
    anadir_hijos(rama1,nodos_iniciales1, tomas_pendientes_dia1,1,6)
    
    rama_dia_1.append([dia1,rama1[0]])

  return rama_dia_1

def calcula_ramas_dia2(rama_dia_1,tomas_pendientes):
  rama_dia_2 = []
  for elemento in rama_dia_1:
    dia1 = elemento[0]
    for dia2 in elemento[1]:
      #print("dia1:", dia1)
      tomas_pendientes_dia2 = tomas_pendientes - dia1 - dia2
      toma_mas_actores2 = toma_con_mas_actores(TOMAS, tomas_pendientes_dia2)
      nodo_inicial2 = {toma_mas_actores2}
      nodos_iniciales2 = []
      nodos_iniciales2.append(nodo_inicial2)
      rama2 = []
      anadir_hijos(rama2,nodos_iniciales2, tomas_pendientes_dia2,1,6)
      
      rama_dia_2.append([dia1,dia2,rama2[0]])

  return rama_dia_2

def calcula_ramas_dia3(rama_dia_2,tomas_pendientes):
  rama_dia_3 = []
  for elemento in rama_dia_2:
    dia1 = elemento[0]
    dia2 = elemento[1]
    for dia3 in elemento[2]:
      tomas_pendientes_dia3 = tomas_pendientes - dia1 - dia2 - dia3
      toma_mas_actores3 = toma_con_mas_actores(TOMAS, tomas_pendientes_dia3)

      nodo_inicial3 = {toma_mas_actores3}
      nodos_iniciales3 = []
      nodos_iniciales3.append(nodo_inicial3)
      rama3 = []
      anadir_hijos(rama3,nodos_iniciales3, tomas_pendientes_dia3,1,6)
      
      rama_dia_3.append([dia1,dia2,dia3, rama3[0]])

  return rama_dia_3

def calcula_ramas_dia4(rama_dia_3,tomas_pendientes):
  rama_dia_4 = []
  for elemento in rama_dia_3:
    dia1 = elemento[0]
    dia2 = elemento[1]
    dia3 = elemento[2]
    for dia4 in elemento[3]:
      tomas_pendientes_dia4 = tomas_pendientes - dia1 - dia2 - dia3 - dia4
      toma_mas_actores4 = toma_con_mas_actores(TOMAS, tomas_pendientes_dia4)

      nodo_inicial4 = {toma_mas_actores4}
      nodos_iniciales4 = []
      nodos_iniciales4.append(nodo_inicial4)
      rama4 = []
      anadir_hijos(rama4,nodos_iniciales4, tomas_pendientes_dia4,1,6)
      
      rama_dia_4.append([dia1,dia2,dia3,dia4, rama4[0]])

  return rama_dia_4

def calcula_mejor_solucion(TOMAS, soluciones):
  mejor_solucion = list()
  coste_mejor_solucion = 10e10

  for solucion in soluciones:  
    coste = coste_de_solucion(TOMAS,solucion,1)
    
    if (coste < coste_mejor_solucion):
        print("solución parcial: ", solucion)
        print("coste parcial:", coste)
        mejor_solucion = solucion 
        coste_mejor_solucion = coste         
  
  print("Mejor solucion:", mejor_solucion)
  print("Coste mejor solucion:", coste_mejor_solucion)
  
tomas_totales = set(range(1,len(TOMAS)+1)) 
ramificacion_y_poda(TOMAS,tomas_totales)


Nivel complejidad. Total soluciones:  19874
solución parcial:  [{1, 2, 20, 6, 7, 22}, {17, 3, 4, 19, 11, 15}, {18, 8, 9, 10, 12, 14}, {16, 5, 25, 28, 13, 30}, {21, 23, 24, 26, 27, 29}]
coste parcial: 30
solución parcial:  [{1, 2, 20, 6, 7, 22}, {17, 3, 4, 19, 11, 15}, {18, 23, 8, 9, 12, 14}, {13, 26, 24, 10, 27, 29}, {16, 5, 21, 25, 28, 30}]
coste parcial: 29
solución parcial:  [{1, 2, 20, 6, 7, 22}, {26, 5, 8, 10, 11, 12}, {3, 4, 21, 9, 29, 15}, {16, 25, 27, 28, 13, 30}, {17, 18, 19, 23, 24, 14}]
coste parcial: 28
solución parcial:  [{1, 2, 20, 6, 7, 22}, {26, 5, 9, 10, 11, 12}, {3, 4, 21, 8, 29, 15}, {16, 25, 27, 28, 13, 30}, {17, 18, 19, 23, 24, 14}]
coste parcial: 27
Mejor solucion: [{1, 2, 20, 6, 7, 22}, {26, 5, 9, 10, 11, 12}, {3, 4, 21, 8, 29, 15}, {16, 25, 27, 28, 13, 30}, {17, 18, 19, 23, 24, 14}]
Coste mejor solucion: 27
Duración:  15  segundos


( * ) ***Calcula la complejidad del algoritmo ***

Respuesta
Mínimo: puedo generar cada nodo con una sola hoja. Es decir, tengo tomas que por la participación de los actores se ahorra y coincide con que se realicen ese día. Es muy poco probable.<br> 

Máximo: por cada día: elijo 1 toma, quedan 5 tomas que pueden tener como máximo 5 opciones para cada día. Podría ser muy elevado pero también es muy poco probable.<br>

Los resultados tienen a ser bajos. En el caso del ejemplo **19874**.

***Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios***

Respuesta

In [0]:
import random
TOMAS_RANDOM=[[random.randint(0,1) for x in range(10)] for y in range(30)]

***Aplica el algoritmo al juego de datos generado***

Respuesta

In [16]:
print(TOMAS_RANDOM)
tomas_totales = set(range(1,len(TOMAS_RANDOM)+1)) 
ramificacion_y_poda(TOMAS_RANDOM,tomas_totales)

[[0, 0, 0, 1, 0, 0, 0, 0, 1, 1], [1, 0, 0, 0, 1, 1, 1, 0, 0, 1], [1, 0, 0, 0, 0, 1, 1, 0, 0, 0], [0, 0, 1, 0, 0, 1, 0, 1, 1, 1], [0, 0, 0, 1, 0, 0, 0, 1, 1, 1], [1, 1, 1, 0, 0, 0, 0, 0, 1, 0], [1, 1, 0, 1, 1, 0, 0, 0, 1, 1], [1, 0, 1, 0, 0, 0, 1, 1, 1, 1], [0, 0, 1, 1, 1, 1, 0, 0, 0, 0], [0, 0, 1, 1, 0, 0, 1, 1, 1, 0], [0, 1, 0, 1, 0, 0, 0, 0, 0, 1], [1, 0, 1, 0, 1, 1, 0, 0, 0, 0], [0, 1, 0, 0, 0, 0, 0, 1, 1, 0], [0, 1, 0, 1, 0, 1, 1, 1, 1, 0], [0, 0, 0, 0, 0, 1, 1, 0, 1, 1], [0, 0, 1, 0, 0, 1, 1, 0, 1, 0], [1, 0, 0, 1, 0, 0, 1, 1, 1, 1], [0, 0, 0, 0, 1, 0, 1, 0, 0, 0], [1, 0, 1, 1, 0, 0, 0, 0, 1, 0], [1, 1, 0, 1, 0, 0, 1, 1, 1, 1], [0, 0, 1, 1, 1, 1, 0, 1, 0, 1], [1, 1, 1, 0, 0, 0, 1, 1, 0, 0], [1, 1, 0, 0, 1, 0, 0, 1, 1, 1], [1, 1, 1, 0, 0, 0, 0, 1, 0, 1], [0, 0, 0, 0, 0, 0, 0, 0, 1, 1], [1, 1, 1, 0, 0, 0, 1, 1, 0, 0], [1, 0, 1, 0, 1, 1, 1, 1, 0, 1], [1, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 1, 0, 1, 0, 1, 0, 1, 0, 0], [1, 1, 1, 0, 1, 1, 0, 1, 0, 0]]
Nivel complejidad. Total soluciones:  5

***Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo***

**Respuesta**<br>
Principalmente he utilizado **Actividad Guiada 2**

Lectura rápida de:<br>
https://www.researchgate.net/publication/320427137_Algoritmo_ACO_aplicado_al_TSP_Resumen_de_una_experiencia_practica <br>
https://imarrero.webs.ull.es/sctm03.v2/modulo1/JJSalazar.pdf <br>
https://www.iit.comillas.edu/aramos/presentaciones/t_mmo_M.pdf <br>
http://bibing.us.es/proyectos/abreproy/5760/fichero/PFC_Jesus_Vazquez.pdf <br>
http://www.cs.us.es/~fsancho/?e=71<br>




***Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño***

**Respuesta**<br>
Incluir costes de los actores<br>
Probar con la función heurística variando determinados parámetros:
 + método de selección de la primera toma de cada día
 + si sólo queda una toma un día: no seleccionar una toma si todos los actores son nuevos y pasarla al día siguiente (no hay restricción de tiempo)

Implementar búsquedas locales con generador de vecindad asociado a la "toma" con más actores comunes <br>


